### DISCUSSION TITLE
如何在使用zjuthesis写作个人论文时依然能正确运行CI

### DISCUSSION THREAD

**Discussion by MARMOTatZJU** [TheNetAdmin/zjuthesis#118] [General]
使用zjuthesis时，如果直接修改默认的zjuthesis.tex及其调用的tex脚本，在CI期间，编译得到的pdf应该会和原先的不一致而失败（如下图所示）。如果依然希望在此时使用CI来验证工程的正确性，请问比较好的实践是？

![image](https://user-images.githubusercontent.com/32362315/103499433-ffd81b80-4e82-11eb-859e-d6475fa391fd.png)

比如另外创建入口personal-thesis.tex然后修改latexmkrc的内容（@default_files=('zjuthesis.tex')）？

**Comment by TheNetAdmin** [comment]
你可以把相关的 check 代码注释掉，暂时想不到其他很好的办法

**Reply by TheNetAdmin** [reply]
我以后有空更新一下 CI，使其每次使用固定的生成的内容，而不是直接使用body文件夹下的内容

**Comment by muzimuzhi** [comment]
可以设置触发 CI 的 branch 白名单和黑名单，例如以下设置了所有以 `-thesis` 结尾的 branch，都不触发。示例见 https://github.com/muzimuzhi/zjuthesis/commit/0efeb7c2d72dfb4217c2ceeda39cc6c6b842a5af

更新：更复杂的条件也可以设置（挺好玩的），不过如果用户没有修改模板本身，写作周期也没有长到因为 latex 组件更新而引入新问题（还要保证本地和 CI 里的版本一致……），基本都不用触发 CI。

文档：[Workflow syntax for GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestbranchestags)

```diff
diff --git a/.github/workflows/build_test.yml b/.github/workflows/build_test.yml
index 432145b..7978232 100644
--- a/.github/workflows/build_test.yml
+++ b/.github/workflows/build_test.yml
@@ -2,6 +2,8 @@ name: Build Tests

 on:
   push:
+    branches-ignore:
+      - "*-thesis"
     paths-ignore:
       - ".github/**"
       - ".gitattributes"
```

**Reply by TheNetAdmin** [reply]
是个好办法，另一种可能的办法是只在我自己的repo中检查pdf在其他用户的repo中不检查

**Reply by muzimuzhi** [reply]
实践上没问题，从开源角度，这稍微阻碍了他人在 fork 里（二次）开发。

另一个不需要修改代码的办法是，在 fork 的 Settings -> Action -> Actions permissions 里，设置不触发 github action ([ref](https://github.community/t/stop-github-actions-running-on-a-fork/17965/2))

PS：感觉可以开启 wiki，把一些有用的信息逐渐整合进去。现在主要依靠手动 pin issue，例如 #94。

**Reply by TheNetAdmin** [reply]
那可能还有办法就是，在使用 PDF 测试样式时，使用自动生成的一些正文内容，而不是使用 body 下的用户内容。

这样大家都可以用，而且以后主模板更新样例正文也不受影响（因为 PDF 测试使用的是额外生成的正文内容）

**Reply by MARMOTatZJU** [reply]
嗯嗯。理论上应该规约一些使用方法（比如推荐创建projects文件夹，在projects下管理自己的编译入口/latexmkrc等），做到用户代码和模版代码的隔离。

这样也能方便用户将自己在模版中做出的改进/适配来pick-cherry到主线分支上来，做出contribution。

**Reply by MARMOTatZJU** [reply]
@TheNetAdmin 用户的repo中做CI检查应该是有好处的，这样用户能知道自己对模版的改进是否会有潜在的问题。

关键可能还是要尽量实现模版的代码和用户自己的代码的隔离。
